[Docu 2470] Configure Rate Limiting (#4270)
Kong/docs.konghq.com: src/gateway/get-started/configure-ratelimiting.md

---
title: Protect your Services
content-type: tutorial
---

## What is rate limiting?

> Rate limiting is used to control the rate of requests sent to an upstream service. It can be used to prevent DoS attacks, limit web scraping, and other forms of overuse. Without rate limiting, a user may make request as often as they like, this leads to traffic spikes and it can impact other users of your upstream services. In {{site.base_gateway}} rate limiting allows you to set parameters that limit requests to your upstream service. 

Reviewer: ```suggestion
Rate limiting is used to control the rate of requests sent to an upstream service. It can be used to prevent DoS attacks, limit web scraping, and other forms of overuse. Without rate limiting, a user may make requests as often as they like, this leads to traffic spikes and it can impact other users of your upstream services. In {{site.base_gateway}} rate limiting allows you to set parameters that limit requests to your upstream service. 
```

## The rate limiting plugin

{{site.base_gateway}} manages rate limiting through the use of Kong's [Rate Limiting plugin](/hub/kong-inc/rate-limiting). The rate limiting plugin has an open source and an enterprise version, with the enterprise version giving you access to features like [sliding window algorithm support](https://en.wikipedia.org/wiki/Sliding_window_protocol), and Redis support. 

> The rate limiting plug-ins limit how often each user can call the API. This protects them from inadvertent or malicious overuse. Without rate limiting, each user may request as often as they like, which can lead to “spikes” of requests that starve other consumers. After rate limiting is enabled, they are limited to a fixed number of requests per second. Kong offers an open source and Enterprise version of the rate limiting plug-in with the Enterprise version providing support for the sliding window algorithm to prevent the API from being overloaded near window boundaries and adds Redis support for greater performance. For this guide we will use the Enterprise version of plug-in.  More details on the advanced rate limiting plug-in can be found here.

Reviewer: ```suggestion
The rate limiting plug-ins limit how often each user can call the API. This protects them from inadvertent or malicious overuse. Without rate limiting, each user may request as often as they like, which can lead to “spikes” of requests that starve other consumers. After rate limiting is enabled, they are limited to a fixed number of requests per second. Kong offers an open source and an Enterprise version of the rate limiting plug-in, with the Enterprise version providing support for the sliding window algorithm to prevent the API from being overloaded near window boundaries, and adds Redis support for greater performance. For this guide we will use the Enterprise version of plug-in.  More details on the advanced rate limiting plug-in can be found here.
```

## Configure the Rate Limiting plugin

To activate a plugin send a `POST` request to the [plugins](/gateway/latest/admin-api/#add-plugin) object of the Admin API: 

```sh
curl -i -X POST http://localhost:8001/plugins \
  --data name=rate-limiting \
  --data config.minute=5 \
  --data config.policy=local
```

This request configures the rate limiting plugin to enforce rate limiting on requests that exceed 5 per minute.

### validate rate limiting

After configuring rate limiting, you can verify that it was configured correctly and is working, by creating 6 requests within a 5 minute period. 
To validate rate limiting, send a request to the API six (6) times from the CLI to confirm the requests are rate limited. 
If you configured {{site.base_gateway}} with the [configure services and routes](/gateway/latest/get-started/configure-services-and-routes) guide use the example below, otherwise substitute the existing values for your own: 

{% navtabs %}
{% navtab Admin API %}

```sh
curl -i -X GET http://localhost:8000/mock/request
```

{% endnavtab %}
{% navtab Web browser %}

> Or with follow these instructions from your web browser:

Reviewer: ```suggestion
Or you can follow these instructions from your web browser:
```
